File: python_problems/KISSING2.py
from s2mpjlib import *
class  KISSING2(CUTEst_problem):

#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
# 
# 
#    Problem: A second formulation of the KISSING NUMBER PROBLEM
#                                                                    
#    Source: This problem is associated to the family of Hard-Spheres 
#    problem. It belongs to the family of sphere packing problems, a 
#    class of challenging problems dating from the beginning of the 
#    17th century which is related to practical problems in Chemistry, 
#    Biology and Physics. Given a fixed unit sphere at the origin in R^n, 
#    the problem consists of arranging a further m unit spheres so that 
#    sum of the distances to these spheres is as small as possible.
#    This problem may be reduced to a nonconvex nonlinear optimization 
#    problem with a potentially large number of (nonoptimal) points 
#    satisfying optimality conditions. We have, thus, a class of problems 
#    indexed by the parameters m and n, that provides a suitable 
#    set of test problems for evaluating nonlinear programming codes.
#    After some algebric manipulations, we can formulate this problem as
#               m
#     Minimize sum <p_i,p_i> - m n
#              i=1
#     subject to
#        
#      <p_i - p_j, p_i - p_j> >= 4 for all different pair of indices i, j
#      and  
#      <p_i, p_i> >= 4 for all indices i
#   
#      as well as n(n-1)/2 normalisation constraints fixing components.
#      The goal is to find an objective value equal to 0.
#      [1]  "Sphere Packings, Lattices and Groups", J. H. Conway and 
#            N. J. C. Sloane, Springer-Verlag, NY, 1988.
#    SIF input: Nick Gould, September 2000
# 
#    classification = "C-CQQR2-RN-V-V"
# 
# **********************************************************************
# 
#    Number of points: m
# 
#           Alternative values for the SIF file parameters:
# IE m                   24             $-PARAMETER  number of points
# IE m                   25             $-PARAMETER  number of points
# 
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#   Translated to Python by S2MPJ version 31 X 2025
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

    name = 'KISSING2'

    def __init__(self, *args): 
        import numpy as np
        from scipy.sparse import csr_matrix
        nargin   = len(args)

        #%%%%%%%%%%%%%%%%%%%  PREAMBLE %%%%%%%%%%%%%%%%%%%%
        v_  = {}
        ix_ = {}
        ig_ = {}
        if nargin<1:
            v_['m'] = int(25);  #  SIF file default value
        else:
            v_['m'] = int(args[0])
# IE m                   100            $-PARAMETER  number of points
# IE n                    4             $-PARAMETER  dimension of sphere
        if nargin<2:
            v_['n'] = int(4);  #  SIF file default value
        else:
            v_['n'] = int(args[1])
# IE n                    8             $-PARAMETER  dimension of sphere
        v_['1'] = 1
        v_['2'] = 2
        v_['n-1'] = v_['n']-v_['1']
        v_['rm'] = float(v_['m'])
        v_['rn'] = float(v_['n'])
        v_['RM+N'] = v_['rm']+v_['rn']
        v_['mn'] = v_['rm']*v_['rn']
        v_['PI/4'] = np.arctan(1.0)
        v_['PI'] = 4.0*v_['PI/4']
        v_['PI/m'] = v_['PI']/v_['rm']
        v_['2PI/m'] = 2.0*v_['PI/m']
        #%%%%%%%%%%%%%%%%%%%  VARIABLES %%%%%%%%%%%%%%%%%%%%
        self.xnames = np.array([])
        self.xscale = np.array([])
        intvars   = np.array([])
        binvars   = np.array([])
        irA          = np.array([],dtype=int)
        icA          = np.array([],dtype=int)
        valA         = np.array([],dtype=float)
        for I in range(int(v_['1']),int(v_['m'])+1):
            for J in range(int(v_['1']),int(v_['n'])+1):
                [iv,ix_,_] = s2mpj_ii('P'+str(I)+','+str(J),ix_)
                self.xnames=arrset(self.xnames,iv,'P'+str(I)+','+str(J))
        #%%%%%%%%%%%%%%%%%%  DATA GROUPS %%%%%%%%%%%%%%%%%%%
        self.gscale  = np.array([])
        self.grnames = np.array([])
        cnames       = np.array([])
        self.cnames  = np.array([])
        gtype        = np.array([])
        [ig,ig_,_] = s2mpj_ii('OBJ',ig_)
        gtype = arrset(gtype,ig,'<>')
        for I in range(int(v_['1']),int(v_['m'])+1):
            for J in range(int(v_['1']),int(v_['m'])+1):
                [ig,ig_,_] = s2mpj_ii('C'+str(I)+','+str(J),ig_)
                gtype = arrset(gtype,ig,'>=')
                cnames = arrset(cnames,ig,'C'+str(I)+','+str(J))
        #%%%%%%%%%%%%%% GLOBAL DIMENSIONS %%%%%%%%%%%%%%%%%
        self.n   = len(ix_)
        ngrp   = len(ig_)
        legrps = np.where(gtype=='<=')[0]
        eqgrps = np.where(gtype=='==')[0]
        gegrps = np.where(gtype=='>=')[0]
        self.nle = len(legrps)
        self.neq = len(eqgrps)
        self.nge = len(gegrps)
        self.m   = self.nle+self.neq+self.nge
        self.congrps = np.concatenate((legrps,eqgrps,gegrps))
        self.cnames = cnames[self.congrps]
        self.nob = ngrp-self.m
        self.objgrps = np.where(gtype=='<>')[0]
        #%%%%%%%%%%%%%%%%%% CONSTANTS %%%%%%%%%%%%%%%%%%%%%
        self.gconst = np.zeros((ngrp,1))
        self.gconst = arrset(self.gconst,ig_['OBJ'],float(v_['mn']))
        for I in range(int(v_['1']),int(v_['m'])+1):
            for J in range(int(v_['1']),int(v_['m'])+1):
                self.gconst = arrset(self.gconst,ig_['C'+str(I)+','+str(J)],float(4.0))
        #%%%%%%%%%%%%%%%%%%%%  BOUNDS %%%%%%%%%%%%%%%%%%%%%
        self.xlower = np.zeros((self.n,1))
        self.xupper = np.full((self.n,1),float('inf'))
        for I in range(int(v_['1']),int(v_['m'])+1):
            for J in range(int(v_['1']),int(v_['n'])+1):
                self.xlower[ix_['P'+str(I)+','+str(J)]] = -float('Inf')
                self.xupper[ix_['P'+str(I)+','+str(J)]] = +float('Inf')
        for I in range(int(v_['2']),int(v_['n'])+1):
            for J in range(int(I),int(v_['n'])+1):
                self.xlower[ix_['P'+str(I)+','+str(J)]] = 0.0
                self.xupper[ix_['P'+str(I)+','+str(J)]] = 0.0
        #%%%%%%%%%%%%%%%%%%% START POINT %%%%%%%%%%%%%%%%%%
        self.x0 = np.zeros((self.n,1))
        self.y0 = np.zeros((self.m,1))
        for I in range(int(v_['1']),int(v_['m'])+1):
            v_['RI'] = float(I)
            v_['2PIi/m'] = v_['2PI/m']*v_['RI']
            v_['cos'] = np.cos(v_['2PIi/m'])
            v_['sin'] = np.sin(v_['2PIi/m'])
            v_['cos'] = v_['cos']+v_['cos']
            v_['sin'] = v_['sin']+v_['sin']
            if('P'+str(I)+','+str(int(v_['1'])) in ix_):
                self.x0[ix_['P'+str(I)+','+str(int(v_['1']))]] = float(v_['cos'])
            else:
                self.y0  = (
                      arrset(self.y0,findfirst(self.congrps,lambda x:x==ig_['P'+str(I)+','+str(int(v_['1']))]),float(v_['cos'])))
            for J in range(int(v_['2']),int(v_['n-1'])+1):
                if('P'+str(I)+','+str(J) in ix_):
                    self.x0[ix_['P'+str(I)+','+str(J)]] = float(v_['sin'])
                else:
                    self.y0  = (
                          arrset(self.y0,findfirst(self.congrps,lambda x:x==ig_['P'+str(I)+','+str(J)]),float(v_['sin'])))
        if('P'+str(int(v_['m']))+','+str(int(v_['n'])) in ix_):
            self.x0[ix_['P'+str(int(v_['m']))+','+str(int(v_['n']))]] = float(v_['cos'])
        else:
            self.y0  = (
                  arrset(self.y0,findfirst(self.congrps,lambda x:x==ig_['P'+str(int(v_['m']))+','+str(int(v_['n']))]),float(v_['cos'])))
        #%%%%%%%%%%%%%%%%%%%% ELFTYPE %%%%%%%%%%%%%%%%%%%%%
        iet_  = {}
        elftv = []
        [it,iet_,_] = s2mpj_ii( 'ePROD1', iet_)
        elftv = loaset(elftv,it,0,'P')
        [it,iet_,_] = s2mpj_ii( 'ePROD2', iet_)
        elftv = loaset(elftv,it,0,'Q')
        elftv = loaset(elftv,it,1,'R')
        #%%%%%%%%%%%%%%%%%% ELEMENT USES %%%%%%%%%%%%%%%%%%
        ie_ = {}
        self.elftype = np.array([])
        ielftype     = np.array([])
        self.elvar   = []
        for I in range(int(v_['1']),int(v_['m'])+1):
            v_['I-'] = -1+I
            v_['I+'] = 1+I
            for J in range(int(v_['1']),int(v_['I-'])+1):
                for K in range(int(v_['1']),int(v_['n'])+1):
                    ename = 'E'+str(I)+','+str(J)+','+str(K)
                    [ie,ie_,_] = s2mpj_ii(ename,ie_)
                    self.elftype = arrset(self.elftype,ie,'ePROD2')
                    ielftype = arrset(ielftype,ie,iet_["ePROD2"])
                    vname = 'P'+str(I)+','+str(K)
                    [iv,ix_] = s2mpj_nlx(self,vname,ix_,1,None,None,None)
                    posev = np.where(elftv[ielftype[ie]]=='Q')[0]
                    self.elvar = loaset(self.elvar,ie,posev[0],iv)
                    vname = 'P'+str(J)+','+str(K)
                    [iv,ix_] = s2mpj_nlx(self,vname,ix_,1,None,None,None)
                    posev = np.where(elftv[ielftype[ie]]=='R')[0]
                    self.elvar = loaset(self.elvar,ie,posev[0],iv)
            for K in range(int(v_['1']),int(v_['n'])+1):
                ename = 'E'+str(I)+','+str(I)+','+str(K)
                [ie,ie_,_] = s2mpj_ii(ename,ie_)
                self.elftype = arrset(self.elftype,ie,'ePROD1')
                ielftype = arrset(ielftype,ie,iet_["ePROD1"])
                vname = 'P'+str(I)+','+str(K)
                [iv,ix_] = s2mpj_nlx(self,vname,ix_,1,None,None,None)
                posev = np.where(elftv[ielftype[ie]]=='P')[0]
                self.elvar = loaset(self.elvar,ie,posev[0],iv)
            for J in range(int(v_['I+']),int(v_['m'])+1):
                for K in range(int(v_['1']),int(v_['n'])+1):
                    ename = 'E'+str(I)+','+str(J)+','+str(K)
                    [ie,ie_,_] = s2mpj_ii(ename,ie_)
                    self.elftype = arrset(self.elftype,ie,'ePROD2')
                    ielftype = arrset(ielftype,ie,iet_["ePROD2"])
                    vname = 'P'+str(I)+','+str(K)
                    [iv,ix_] = s2mpj_nlx(self,vname,ix_,1,None,None,None)
                    posev = np.where(elftv[ielftype[ie]]=='Q')[0]
                    self.elvar = loaset(self.elvar,ie,posev[0],iv)
                    vname = 'P'+str(J)+','+str(K)
                    [iv,ix_] = s2mpj_nlx(self,vname,ix_,1,None,None,None)
                    posev = np.where(elftv[ielftype[ie]]=='R')[0]
                    self.elvar = loaset(self.elvar,ie,posev[0],iv)
        #%%%%%%%%%%%%%%%%%%% GROUP USES %%%%%%%%%%%%%%%%%%%
        self.grelt   = []
        for ig in np.arange(0,ngrp):
            self.grelt.append(np.array([]))
        self.grftype = np.array([])
        self.grelw   = []
        nlc         = np.array([])
        for I in range(int(v_['1']),int(v_['m'])+1):
            for K in range(int(v_['1']),int(v_['n'])+1):
                ig = ig_['OBJ']
                posel = len(self.grelt[ig])
                self.grelt  = (
                      loaset(self.grelt,ig,posel,ie_['E'+str(I)+','+str(I)+','+str(K)]))
                nlc = np.union1d(nlc,np.array([ig]))
                self.grelw = loaset(self.grelw,ig,posel,1.)
            for J in range(int(v_['1']),int(v_['m'])+1):
                for K in range(int(v_['1']),int(v_['n'])+1):
                    ig = ig_['C'+str(I)+','+str(J)]
                    posel = len(self.grelt[ig])
                    self.grelt  = (
                          loaset(self.grelt,ig,posel,ie_['E'+str(I)+','+str(J)+','+str(K)]))
                    nlc = np.union1d(nlc,np.array([ig]))
                    self.grelw = loaset(self.grelw,ig,posel,1.)
        #%%%%%%%%%%%%%%%%%% OBJECT BOUNDS %%%%%%%%%%%%%%%%%
#    Solution
# XL SOLUTION             0.00000D+00   $ n=4, m = 24
# XL SOLUTION             6.48030D+00   $ n=4, m = 25 one of many local solutions
        #%%%%%%%% DEFAULT FOR MISSING SECTION(S) %%%%%%%%%%
        #%%%%%%%%%%%%% FORM clower AND cupper %%%%%%%%%%%%%
        self.clower = np.full((self.m,1),-float('Inf'))
        self.cupper = np.full((self.m,1),+float('Inf'))
        self.clower[np.arange(self.nle+self.neq,self.m)] = np.zeros((self.nge,1))
        #%%%% RETURN VALUES FROM THE __INIT__ METHOD %%%%%%
        self.lincons  = (
              np.where(np.isin(self.congrps,np.setdiff1d(self.congrps,nlc)))[0])
        self.pbclass   = "C-CQQR2-RN-V-V"
        self.objderlvl = 2
        self.conderlvl = [2]

# **********************
#  SET UP THE FUNCTION *
#  AND RANGE ROUTINES  *
# **********************

    #%%%%%%%%%%%%%%% NONLINEAR ELEMENTS %%%%%%%%%%%%%%%

    @staticmethod
    def ePROD1(self, nargout,*args):

        import numpy as np
        EV_  = args[0]
        iel_ = args[1]
        f_   = EV_[0]*EV_[0]
        if not isinstance( f_, float ):
            f_   = f_.item();
        if nargout>1:
            try:
                dim = len(IV_)
            except:
                dim = len(EV_)
            g_ = np.zeros(dim)
            g_[0] = EV_[0]+EV_[0]
            if nargout>2:
                H_ = np.zeros((1,1))
                H_[0,0] = 2.0
        if nargout == 1:
            return f_
        elif nargout == 2:
            return f_,g_
        elif nargout == 3:
            return f_,g_,H_

    @staticmethod
    def ePROD2(self, nargout,*args):

        import numpy as np
        EV_  = args[0]
        iel_ = args[1]
        U_ = np.zeros((1,2))
        IV_ = np.zeros(1)
        U_[0,0] = U_[0,0]+1
        U_[0,1] = U_[0,1]-1
        IV_[0] = U_[0:1,:].dot(EV_)
        f_   = IV_[0]*IV_[0]
        if not isinstance( f_, float ):
            f_   = f_.item();
        if nargout>1:
            try:
                dim = len(IV_)
            except:
                dim = len(EV_)
            g_ = np.zeros(dim)
            g_[0] = IV_[0]+IV_[0]
            g_ =  U_.T.dot(g_)
            if nargout>2:
                H_ = np.zeros((1,1))
                H_[0,0] = 2.0
                H_ = U_.T.dot(H_).dot(U_)
        if nargout == 1:
            return f_
        elif nargout == 2:
            return f_,g_
        elif nargout == 3:
            return f_,g_,H_

#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

File: julia_problems/KISSING2.jl
function KISSING2(action::String,args::Union{PBM,Int,Float64,Vector{Int},Vector{Float64}}...)
# 
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
# 
# 
#    Problem: A second formulation of the KISSING NUMBER PROBLEM
#                                                                    
#    Source: This problem is associated to the family of Hard-Spheres 
#    problem. It belongs to the family of sphere packing problems, a 
#    class of challenging problems dating from the beginning of the 
#    17th century which is related to practical problems in Chemistry, 
#    Biology and Physics. Given a fixed unit sphere at the origin in R^n, 
#    the problem consists of arranging a further m unit spheres so that 
#    sum of the distances to these spheres is as small as possible.
#    This problem may be reduced to a nonconvex nonlinear optimization 
#    problem with a potentially large number of (nonoptimal) points 
#    satisfying optimality conditions. We have, thus, a class of problems 
#    indexed by the parameters m and n, that provides a suitable 
#    set of test problems for evaluating nonlinear programming codes.
#    After some algebric manipulations, we can formulate this problem as
#               m
#     Minimize sum <p_i,p_i> - m n
#              i=1
#     subject to
#        
#      <p_i - p_j, p_i - p_j> >= 4 for all different pair of indices i, j
#      and  
#      <p_i, p_i> >= 4 for all indices i
#   
#      as well as n(n-1)/2 normalisation constraints fixing components.
#      The goal is to find an objective value equal to 0.
#      [1]  "Sphere Packings, Lattices and Groups", J. H. Conway and 
#            N. J. C. Sloane, Springer-Verlag, NY, 1988.
#    SIF input: Nick Gould, September 2000
# 
#    classification = "C-QQR2-RN-V-V"
# 
# **********************************************************************
# 
#    Number of points: m
# 
#       Alternative values for the SIF file parameters:
# IE m                   24             $-PARAMETER  number of points
# IE m                   25             $-PARAMETER  number of points
# 
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#   Translated to Julia by S2MPJ version 7 X 2024
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

    name = "KISSING2"

    if action == "setup"
        pb           = PB(name)
        pbm          = PBM(name)
        nargin       = length(args)
        pbm.call     = getfield( Main, Symbol( name ) )

        #%%%%%%%%%%%%%%%%%%%  PREAMBLE %%%%%%%%%%%%%%%%%%%%
        v_  = Dict{String,Float64}();
        ix_ = Dict{String,Int}();
        ig_ = Dict{String,Int}();
        if nargin<1
            v_["m"] = Int64(25);  #  SIF file default value
        else
            v_["m"] = Int64(args[1]);
        end
# IE m                   100            $-PARAMETER  number of points
# IE n                    4             $-PARAMETER  dimension of sphere
        if nargin<2
            v_["n"] = Int64(4);  #  SIF file default value
        else
            v_["n"] = Int64(args[2]);
        end
# IE n                    8             $-PARAMETER  dimension of sphere
        v_["1"] = 1
        v_["2"] = 2
        v_["n-1"] = v_["n"]-v_["1"]
        v_["rm"] = Float64(v_["m"])
        v_["rn"] = Float64(v_["n"])
        v_["RM+N"] = v_["rm"]+v_["rn"]
        v_["mn"] = v_["rm"]*v_["rn"]
        v_["PI/4"] = atan(1.0)
        v_["PI"] = 4.0*v_["PI/4"]
        v_["PI/m"] = v_["PI"]/v_["rm"]
        v_["2PI/m"] = 2.0*v_["PI/m"]
        #%%%%%%%%%%%%%%%%%%%  VARIABLES %%%%%%%%%%%%%%%%%%%%
        pb.xscale = Float64[]
        intvars = Int64[]
        binvars = Int64[]
        for I = Int64(v_["1"]):Int64(v_["m"])
            for J = Int64(v_["1"]):Int64(v_["n"])
                iv,ix_,_ = s2mpj_ii("P"*string(I)*","*string(J),ix_)
                arrset(pb.xnames,iv,"P"*string(I)*","*string(J))
            end
        end
        #%%%%%%%%%%%%%%%%%%  DATA GROUPS %%%%%%%%%%%%%%%%%%%
        gtype    = String[]
        ig,ig_,_ = s2mpj_ii("OBJ",ig_)
        arrset(gtype,ig,"<>")
        for I = Int64(v_["1"]):Int64(v_["m"])
            for J = Int64(v_["1"]):Int64(v_["m"])
                ig,ig_,_ = s2mpj_ii("C"*string(I)*","*string(J),ig_)
                arrset(gtype,ig,">=")
                arrset(pb.cnames,ig,"C"*string(I)*","*string(J))
            end
        end
        #%%%%%%%%%%%%%% GLOBAL DIMENSIONS %%%%%%%%%%%%%%%%%
        pb.n   = length(ix_)
        ngrp   = length(ig_)
        legrps = findall(x->x=="<=",gtype)
        eqgrps = findall(x->x=="==",gtype)
        gegrps = findall(x->x==">=",gtype)
        pb.nle = length(legrps)
        pb.neq = length(eqgrps)
        pb.nge = length(gegrps)
        pb.m   = pb.nle+pb.neq+pb.nge
        pbm.congrps = [[legrps;eqgrps];gegrps]
        pb.nob = ngrp-pb.m
        pbm.objgrps = findall(x->x=="<>",gtype)
        #%%%%%%%%%%%%%%%%%% CONSTANTS %%%%%%%%%%%%%%%%%%%%%
        pbm.gconst = zeros(Float64,ngrp)
        pbm.gconst[ig_["OBJ"]] = Float64(v_["mn"])
        for I = Int64(v_["1"]):Int64(v_["m"])
            for J = Int64(v_["1"]):Int64(v_["m"])
                pbm.gconst[ig_["C"*string(I)*","*string(J)]] = Float64(4.0)
            end
        end
        #%%%%%%%%%%%%%%%%%%%%  BOUNDS %%%%%%%%%%%%%%%%%%%%%
        pb.xlower = zeros(Float64,pb.n)
        pb.xupper =    fill(Inf,pb.n)
        for I = Int64(v_["1"]):Int64(v_["m"])
            for J = Int64(v_["1"]):Int64(v_["n"])
                pb.xlower[ix_["P"*string(I)*","*string(J)]] = -Inf
                pb.xupper[ix_["P"*string(I)*","*string(J)]] = +Inf
            end
        end
        for I = Int64(v_["2"]):Int64(v_["n"])
            for J = Int64(I):Int64(v_["n"])
                pb.xlower[ix_["P"*string(I)*","*string(J)]] = 0.0
                pb.xupper[ix_["P"*string(I)*","*string(J)]] = 0.0
            end
        end
        #%%%%%%%%%%%%%%%%%%% START POINT %%%%%%%%%%%%%%%%%%
        pb.x0 = zeros(Float64,pb.n)
        pb.y0 = zeros(Float64,pb.m)
        for I = Int64(v_["1"]):Int64(v_["m"])
            v_["RI"] = Float64(I)
            v_["2PIi/m"] = v_["2PI/m"]*v_["RI"]
            v_["cos"] = cos(v_["2PIi/m"])
            v_["sin"] = sin(v_["2PIi/m"])
            v_["cos"] = v_["cos"]+v_["cos"]
            v_["sin"] = v_["sin"]+v_["sin"]
            if haskey(ix_,"P"*string(I)*","*string(Int64(v_["1"])))
                pb.x0[ix_["P"*string(I)*","*string(Int64(v_["1"]))]] = Float64(v_["cos"])
            else
                pb.y0[findfirst(x->x==ig_["P"*string(I)*","*string(Int64(v_["1"]))],pbm.congrps)] = Float64(v_["cos"])
            end
            for J = Int64(v_["2"]):Int64(v_["n-1"])
                if haskey(ix_,"P"*string(I)*","*string(J))
                    pb.x0[ix_["P"*string(I)*","*string(J)]] = Float64(v_["sin"])
                else
                    pb.y0[findfirst(x->x==ig_["P"*string(I)*","*string(J)],pbm.congrps)]  = (
                          Float64(v_["sin"]))
                end
            end
        end
        if haskey(ix_,"P"*string(Int64(v_["m"]))*","*string(Int64(v_["n"])))
            pb.x0[ix_["P"*string(Int64(v_["m"]))*","*string(Int64(v_["n"]))]]  = (
                  Float64(v_["cos"]))
        else
            pb.y0[findfirst(x->x==ig_["P"*string(Int64(v_["m"]))*","*string(Int64(v_["n"]))],pbm.congrps)] = Float64(v_["cos"])
        end
        #%%%%%%%%%%%%%%%%%%%% ELFTYPE %%%%%%%%%%%%%%%%%%%%%
        iet_  = Dict{String,Int}()
        elftv = Vector{Vector{String}}()
        it,iet_,_ = s2mpj_ii( "ePROD1", iet_)
        loaset(elftv,it,1,"P")
        it,iet_,_ = s2mpj_ii( "ePROD2", iet_)
        loaset(elftv,it,1,"Q")
        loaset(elftv,it,2,"R")
        #%%%%%%%%%%%%%%%%%% ELEMENT USES %%%%%%%%%%%%%%%%%%
        ie_      = Dict{String,Int}()
        ielftype = Vector{Int64}()
        for I = Int64(v_["1"]):Int64(v_["m"])
            v_["I-"] = -1+I
            v_["I+"] = 1+I
            for J = Int64(v_["1"]):Int64(v_["I-"])
                for K = Int64(v_["1"]):Int64(v_["n"])
                    ename = "E"*string(I)*","*string(J)*","*string(K)
                    ie,ie_,_  = s2mpj_ii(ename,ie_)
                    arrset(pbm.elftype,ie,"ePROD2")
                    arrset(ielftype,ie,iet_["ePROD2"])
                    vname = "P"*string(I)*","*string(K)
                    iv,ix_,pb = s2mpj_nlx(vname,ix_,pb,1,nothing,nothing,nothing)
                    posev = findfirst(x->x=="Q",elftv[ielftype[ie]])
                    loaset(pbm.elvar,ie,posev,iv)
                    vname = "P"*string(J)*","*string(K)
                    iv,ix_,pb = s2mpj_nlx(vname,ix_,pb,1,nothing,nothing,nothing)
                    posev = findfirst(x->x=="R",elftv[ielftype[ie]])
                    loaset(pbm.elvar,ie,posev,iv)
                end
            end
            for K = Int64(v_["1"]):Int64(v_["n"])
                ename = "E"*string(I)*","*string(I)*","*string(K)
                ie,ie_,_  = s2mpj_ii(ename,ie_)
                arrset(pbm.elftype,ie,"ePROD1")
                arrset(ielftype,ie,iet_["ePROD1"])
                vname = "P"*string(I)*","*string(K)
                iv,ix_,pb = s2mpj_nlx(vname,ix_,pb,1,nothing,nothing,nothing)
                posev = findfirst(x->x=="P",elftv[ielftype[ie]])
                loaset(pbm.elvar,ie,posev,iv)
            end
            for J = Int64(v_["I+"]):Int64(v_["m"])
                for K = Int64(v_["1"]):Int64(v_["n"])
                    ename = "E"*string(I)*","*string(J)*","*string(K)
                    ie,ie_,_  = s2mpj_ii(ename,ie_)
                    arrset(pbm.elftype,ie,"ePROD2")
                    arrset(ielftype,ie,iet_["ePROD2"])
                    vname = "P"*string(I)*","*string(K)
                    iv,ix_,pb = s2mpj_nlx(vname,ix_,pb,1,nothing,nothing,nothing)
                    posev = findfirst(x->x=="Q",elftv[ielftype[ie]])
                    loaset(pbm.elvar,ie,posev,iv)
                    vname = "P"*string(J)*","*string(K)
                    iv,ix_,pb = s2mpj_nlx(vname,ix_,pb,1,nothing,nothing,nothing)
                    posev = findfirst(x->x=="R",elftv[ielftype[ie]])
                    loaset(pbm.elvar,ie,posev,iv)
                end
            end
        end
        #%%%%%%%%%%%%%%%%%%% GROUP USES %%%%%%%%%%%%%%%%%%%
        for ig in 1:ngrp
            arrset(pbm.grelt,ig,Int64[])
        end
        nlc = Int64[]
        for I = Int64(v_["1"]):Int64(v_["m"])
            for K = Int64(v_["1"]):Int64(v_["n"])
                ig = ig_["OBJ"]
                posel = length(pbm.grelt[ig])+1
                loaset(pbm.grelt,ig,posel,ie_["E"*string(I)*","*string(I)*","*string(K)])
                arrset(nlc,length(nlc)+1,ig)
                loaset(pbm.grelw,ig,posel,1.)
            end
            for J = Int64(v_["1"]):Int64(v_["m"])
                for K = Int64(v_["1"]):Int64(v_["n"])
                    ig = ig_["C"*string(I)*","*string(J)]
                    posel = length(pbm.grelt[ig])+1
                    loaset(pbm.grelt,ig,posel,ie_["E"*string(I)*","*string(J)*","*string(K)])
                    arrset(nlc,length(nlc)+1,ig)
                    loaset(pbm.grelw,ig,posel,1.)
                end
            end
        end
        #%%%%%%%%%%%%%%%%%% OBJECT BOUNDS %%%%%%%%%%%%%%%%%
#    Solution
# XL SOLUTION             0.00000D+00   $ n=4, m = 24
# XL SOLUTION             6.48030D+00   $ n=4, m = 25 one of many local solutions
        #%%%%%%%% DEFAULT FOR MISSING SECTION(S) %%%%%%%%%%
        #%%%%%%%%%%%%% FORM clower AND cupper %%%%%%%%%%%%%
        pb.clower = -1*fill(Inf,pb.m)
        pb.cupper =    fill(Inf,pb.m)
        pb.clower[pb.nle+pb.neq+1:pb.m] = zeros(Float64,pb.nge)
        pb.cupper[1:pb.nge] = fill(Inf,pb.nge)
        pbm.A = spzeros(Float64,0,0)
        pbm.H = spzeros(Float64,0,0)
        #%%%%% RETURN VALUES FROM THE SETUP ACTION %%%%%%%%
        pb.lincons = findall(x-> x in setdiff( pbm.congrps,nlc),pbm.congrps)
        pb.pbclass = "C-QQR2-RN-V-V"
        pbm.objderlvl = 2
        pb.objderlvl = pbm.objderlvl;
        pbm.conderlvl = [2]
        pb.conderlvl  = pbm.conderlvl;
        return pb, pbm

# **********************
#  SET UP THE FUNCTION *
#  AND RANGE ROUTINES  *
# **********************

    #%%%%%%%%%%%%%%% NONLINEAR ELEMENTS %%%%%%%%%%%%%%%

    elseif action == "ePROD1"

        EV_     = args[1]
        iel_    = args[2]
        nargout = args[3]
        pbm     = args[4]
        f_   = EV_[1]*EV_[1]
        if nargout>1
            dim = try length(IV_) catch; length(EV_) end
            g_  = zeros(Float64,dim)
            g_[1] = EV_[1]+EV_[1]
            if nargout>2
                H_ = zeros(Float64,1,1)
                H_[1,1] = 2.0
            end
        end
        if nargout == 1
            return f_
        elseif nargout == 2
            return f_,g_
        elseif nargout == 3
            return f_,g_,H_
        end

    elseif action == "ePROD2"

        EV_     = args[1]
        iel_    = args[2]
        nargout = args[3]
        pbm     = args[4]
        U_ = zeros(Float64,1,2)
        IV_ =  zeros(Float64,1)
        U_[1,1] = U_[1,1]+1
        U_[1,2] = U_[1,2]-1
        IV_[1] = dot(U_[1,:],EV_)
        f_   = IV_[1]*IV_[1]
        if nargout>1
            dim = try length(IV_) catch; length(EV_) end
            g_  = zeros(Float64,dim)
            g_[1] = IV_[1]+IV_[1]
            g_ =  U_'*g_
            if nargout>2
                H_ = zeros(Float64,1,1)
                H_[1,1] = 2.0
                H_ = U_'*H_*U_
            end
        end
        if nargout == 1
            return f_
        elseif nargout == 2
            return f_,g_
        elseif nargout == 3
            return f_,g_,H_
        end

    #%%%%%%%%%%%%%%% THE MAIN ACTIONS %%%%%%%%%%%%%%%

    elseif action in  ["fx","fgx","fgHx","cx","cJx","cJHx","cIx","cIJx","cIJHx","cIJxv","fHxv",
                       "cJxv","cJtxv","cIJtxv","Lxy","Lgxy","LgHxy","LIxy","LIgxy","LIgHxy",
                       "LHxyv","LIHxyv"]

        pbm = args[1]
        if pbm.name == name
            pbm.has_globs = [0,0]
            return s2mpj_eval(action,args...)
        else
            println("ERROR: please run "*name*" with action = setup")
            return ntuple(i->undef,args[end])
        end

    else
        println("ERROR: action "*action*" unavailable for problem "*name*".jl")
        return ntuple(i->undef,args[end])
    end

end

#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

